### DISCUSSION TITLE
Help with converting PSD file

### DISCUSSION THREAD

**Discussion by clintpike** [ImageMagick/ImageMagick#4143] [Help]
Hi there, 
Our software is trying to convert a PSD file - it runs the below command and RAM on the server hits the max and then the process fails. We have 32GB of RAM on the server and are using IM 6.9.7-4 Q16
Command:
`identify -format %w-@-%h-@-%r-@- path\to\file\file.psd`
The error reported in our log file is:
**Returned code: 1, error output: identify: Memory allocation failed `Cannot allocate memory' @ fatal/string.c/AcquireStringInfo/182._**
The file is 285MB. It has 187 layers in 83 groups and is 1440 x 12704 pixels

Any ideas what is happening here? Why it is trying to use so much RAM? If the server had more, would it work? Is there a way to limit how much it uses?
Many thanks

**Comment by snibgo** [comment]
I'm not sure with your version, but IM may be reading all the pixels into memory. If so, then "-ping" may prevent that.

You can limit memory usage, so pixels are written to disk instead. Do this either with a policy.xml or "-limit" settings in the command line. Disk is slower than memory.

You might identify one image at a time:
```
identify -format %w-@-%h-@-%r-@- path\to\file\file.psd[0]
```
Then put that in a loop so you identify [1] then [2] etc.

**Reply by clintpike** [reply]
Thanks for the info about identity.

The issue more is with running convert on the same PSD file - we are trying to produce a jpg from the psd.

Just to recap, the file is 285MB. It has 187 layers in 83 groups and is 1440 x 12704 pixels

We are running a simple command on the command line:
`convert path\to\file\file.psd converted_file.jpg`

Running this on a box with 32GB RAM and the RAM just maxes out and then fails:
**internal buffer error : Memory allocation failed : growing buffer
internal buffer error : Memory allocation failed : growing buffer
convert: Memory allocation failed `Not enough space' @ fatal/string.c/AcquireStringInfo/196.**

Tried running this on the latest version `6.9.12-21 Q16` (the latest version of 6.9) and version `7.0.8-49 Q16` with exactly the same results.

How do you use the -limit parameter? We have tried using it on the command line but it does not make a difference:
`convert path\to\file\file.psd -limit memory 5GiB -limit map 5GiB converted_file.jpg`

Why is IM demanding so much RAM? Is this by design? If we have 100GB - would it use all of that as well?

Many thanks

**Comment by snibgo** [comment]
I don't use Photoshop, and I'm not sure what "187 layers in 83 groups" means. How many images can ImageMagick see in the file?

If IM can see 187 images, there are a total of 187\*1440\*12704 = 3.4 billion pixels. If you try to read them all, this would exceed the available memory.

You say:
> ... we are trying to produce a jpg from the psd.

Do you want one jpeg, or 187 jpegs? If you want just one, this would probably be from just the first image in the PSD:
```
convert path\to\file\file.psd[0] converted_file.jpg
```
Note the [0], which means "read just the first image". Your command, without [0], will try to read all the images, and make a jpeg from each image it finds.

If you want one jpeg from each image, you can run separate commads, such as:
```
convert path\to\file\file.psd[0] converted_file_000.jpg
convert path\to\file\file.psd[1] converted_file_001.jpg
convert path\to\file\file.psd[2] converted_file_002.jpg
:
:
```
... or use a loop. What shell do you use? Bash, Windows or what?


> Why is IM demanding so much RAM? Is this by design? If we have 100GB - would it use all of that as well?

Yes, it is by design. The command you showed tells IM to read all the pixels in all the images. The memory required per pixel depends on your version of IM. What is the complete text output from "convert -version"? Q16 v6 without HDRI needs 8 bytes/pixel, so that would be 27 GB.

JPEGs are only 8 bits per channel per pixel, so you could probably use IM Q8, which would halve the memory usage.
